tests/devices/default_qubit/test_default_qubit.py

```diff
@@ -317,6 +317,17 @@ def circuit(p):
         assert qml.math.get_interface(res) == "tensorflow"
         assert qml.math.allclose(res, -1)
 
+    def test_basis_state_wire_order(self):
+        """Test that the wire order is correct with a basis state if the tape wires have a non standard order."""
+
+        dev = DefaultQubit()
+
+        tape = qml.tape.QuantumScript([qml.BasisState([1], wires=1), qml.PauliZ(0)], [qml.state()])
+
+        expected = np.array([0, 1, 0, 0], dtype=np.complex128)
+        res = dev.execute(tape)
+        assert qml.math.allclose(res, expected)
+
```

Comment:
Do we also have tests for custom wire labels? E.g., strings or just non-consecutive integers?

Reply:
Good catch!

```
from pennylane.devices import DefaultQubit

dev = DefaultQubit()

tape = qml.tape.QuantumScript([qml.BasisState([1], wires="A"), qml.PauliZ("C")], [qml.state()])

expected = np.array([0, 1, 0, 0], dtype=np.complex128)
res = dev.execute(tape)
```

```
>>> res
array([ 0.+0.j, -0.+0.j,  1.+0.j, -0.+0.j])
```

Reply:
So this same kind of problem won't occur in those cases. If the tape wires are `(2,0)` or `("b", "a")`, then they will get mapped to `(0,1)` at the beginning.

The problem is that the wire order `(1,0)` is not mapped to `(0,1)`, but left as is.

Reply:
@isaacdevlugt  I'm a little confused as to why the "expected" is `np.array([0, 1, 0, 0], dtype=np.complex128)` in this case.


The above example is equivalent to:
```
>>> tape = qml.tape.QuantumScript([qml.BasisState([1], wires=0), qml.PauliZ(1)], [qml.state()])
>>> dev.execute(tape)

array([ 0.+0.j, -0.+0.j,  1.+0.j, -0.+0.j])
```

which is exactly the same.